### PR TITLE
Support relative map size

### PR DESCRIPTION
**Pull request by Pessimistress**
For #561 

### Change list
- Accept CSS strings for width/height props
- Add `onResize` callback to `StaticMap` and `InteractiveMap`

Tested DOM/map/overlay synchronization with the map control and deck.gl examples

@TrySound

### CODE REVIEW

File: src/components/static-map.js [hunk 1]
```diff
@@ -98,22 +118,8 @@ export default class StaticMap extends PureComponent {
...
-      width: this.props.width,
-      height: this.props.height
-    });
-  }
-
-  componentDidUpdate() {
-    // Since Mapbox's map.resize() reads size from DOM
-    // we must wait to read size until after render (i.e. here in "didUpdate")
-    this._updateMapSize(this.state, this.props);
+    this._updateMapProps(newProps);
```
Comment (TrySound) [inline src/components/static-map.js]: Can this be moved to `componentDidUpdate`?

File: src/components/static-map.js [hunk 2]
```diff
@@ -192,51 +207,59 @@ export default class StaticMap extends PureComponent {
...
+          key: 'map-mapbox',
+          ref: this._mapboxMapLoaded,
+          style: mapStyle,
+          className
+        }),
+        createElement(AutoSizer, {
+          key: 'autosizer',
+          disableWidth: Number.isFinite(width),
+          disableHeight: Number.isFinite(height)
+        }, this._renderOverlays.bind(this)),
```
Comment (TrySound) [inline src/components/static-map.js]: I would do this in constructor or just use class properties.

Comment (Pessimistress) [reply src/components/static-map.js]: This is subtle... We need `AutoSizer` to rerender when either the props of `StaticMap` change or the container is resized. But it's a pure component. `bind` makes a new child every render pass and triggers that rerender.

File: src/mapbox/mapbox.js [hunk 3]
```diff
@@ -250,6 +254,14 @@ export default class Mapbox {
 
     this._create(props);
 
+    // Hijack dimension properties
+    // This eliminates the timing issue between calling resize() and DOM update
+    const {container} = props;
+    container.__defineGetter__('offsetWidth', () => this.width);
+    container.__defineGetter__('clientWidth', () => this.width);
+    container.__defineGetter__('offsetHeight', () => this.height);
+    container.__defineGetter__('clientHeight', () => this.height);
```
Comment (schnerd) [inline src/mapbox/mapbox.js]: Clever. 

Would Object.defineProperty be better here? Sounds like defineGetter is deprecated.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/__defineGetter__

File: test/src/components/map.spec.js [hunk 4]
```diff
@@ -155,7 +155,8 @@ TEST_CASES.forEach(testCase => {
   });
 });
 
-test('Interactive map renders children on first render', t => {
+// children are no longer rendered on first pass due to auto sizer
+test.skip('Interactive map renders children on first render', t => {
```
Comment (schnerd) [inline test/src/components/map.spec.js]: Might as well delete the test?

**Comment by wafisher** [comment]
How can I have the component fill its space and _keep_ filling its space when resized? I could attach a function to `resize` but what would it do, set `height: '100%', width: '100%'` again, which wouldn't cause a re-render.

**Comment by wafisher** [comment]
Just solved my issue. The problem was this code:

```
<ReactMapGL
  {...this.state.viewport}
  onViewportChange={viewport => this.setState({ viewport })}
  [...other props...]
/>
```

The `onViewportChange` would give real numbers (not "100%") for width and height, which would then set the component with those numbers, not CSS styles. I changed by keeping width/height out of viewport and setting explicitly on `ReactMapGL` component.

**Comment by invisibleroads** [comment]
Here is how we solved it, thanks to @wafisher's comment that `onViewportChange` tries to explicitly set width and height. 

```
class Map extends Component {
  state = { 
    viewport: {
    },
  } 

  onViewportChange = viewport => { 
    const {width, height, ...etc} = viewport
    this.setState({viewport: etc})
  } 

  render () { 
    const { 
      viewport,
    } = this.state
    return ( 
      <ReactMapGL
        width='100%'
        height='100%'
        {...viewport}
        mapboxApiAccessToken={process.env.REACT_APP_MAPBOX_TOKEN}
        onViewportChange={viewport => this.onViewportChange(viewport)}
      />
    ) 
  } 
}
```
@salah93

**Pull request merged**